subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -198,4 +204,25 @@ public Configuration getCompileDependencies() {
             return convention.getProject().getConfigurations().getByName(JavaPlugin.COMPILE_CONFIGURATION_NAME);
         }
     }
+
+    private class SourceSetOutputCleanUpRegistrationListener implements ProjectEvaluationListener {
```

Comment:
Now that I've seen how this fits with other things.. I'm wondering if we should make all of these ModelConfigurationListeners so that we don't have to worry about a different afterEvaluate adding more configuration later (ignoring the fact we also have configure tasks...).